fs: add littlefs as a package

vincent-d opened the pull request:
Just to avoid duplicate work, I'm pushing this. I started to work on littlefs package integration. I wanted to test it and see the performances.
I still have some issues and I don't know when I'll have time to continue working on it.

Fix #8204

vincent-d commented:
Looks to work now (at least unittests do not crash anymore ;)
If someone wants to play with it, it's possible as is. It would need some more clean up if we want to integrate it.

I found out that littlefs uses `malloc`, there is an option to not use it, but then only one file can be opened at a time...

vincent-d commented:
Removed the WIP. Should be ready for review.

smlng commented on pkg/littlefs/fs/littlefs_fs.c:
```diff
@@ -0,0 +1,499 @@
+static int _dev_read(const struct lfs_config *c, lfs_block_t block,
+                 lfs_off_t off, void *buffer, lfs_size_t size)
+{
+    (void)block;
```
why, its used in DEBUG and mtd_read

vincent-d replied:
Removed

smlng commented on pkg/littlefs/fs/littlefs_fs.c:
```diff
@@ -0,0 +1,499 @@
+static int _dev_write(const struct lfs_config *c, lfs_block_t block,
+                  lfs_off_t off, const void *buffer, lfs_size_t size)
+{
+    (void)block;
```
dito

vincent-d replied:
Removed

smlng commented on pkg/littlefs/fs/littlefs_fs.c:
```diff
@@ -0,0 +1,499 @@
+    if (ret >= 0) {
+        return 0;
+    }
+    else {
```
else not needed, due to return above

vincent-d replied:
Removed

smlng commented on sys/include/fs/littlefs_fs.h:
```diff
@@ -0,0 +1,105 @@
+ * @{
+ */
+#ifndef LITTLEFS_LOOKAHED_SIZE
+/** Default lookahed size */
```
typo -> `look ahead`?

vincent-d replied:
Fixed, here and in variable names

smlng commented on cpu/native/mtd/mtd_native.c:
```diff
@@ -37,7 +37,7 @@ static int _init(mtd_dev_t *dev)
 
     if (!f) {
-        DEBUG("mtd_native: init: creating file %s\n", name);
+        DEBUG("mtd_native: init: creating file %s\n", _dev->fname);
```
nice catch!

vincent-d replied:
The credit goes to @MichelRottleuthner (this comes from #7104) ;)

smlng replied:
ah, so this is kind of based on #7104? and we should merge that first?

vincent-d replied:
This is not really based on #7104, I simply copied this fix and the `vfs.h` part as well.
So we can merge independently, I guess only a rebase will be needed.

smlng commented on pkg/littlefs/Makefile.littlefs:
```diff
@@ -0,0 +1,5 @@
+MODULE := littlefs
+
+CFLAGS += -Wno-unused-parameter -Wno-format
```
mhm, might be worth upstreaming some fixing then 😉

vincent-d replied:
Yeah, we'll see later ^^

smlng commented on pkg/littlefs/fs/littlefs_fs.c:
```diff
@@ -0,0 +1,499 @@
+ */
+
+/**
+ * @ingroup     fs
```
should be `pkg_littlefs`?  Btw. that group is not defined, as far as I can see, right? should add a `doc.txt` to do so.

vincent-d replied:
True, but actually C files are not parsed by Doxygen.
There is already a `doc.txt` in the pkg directory defining the groups.

smlng commented on pkg/littlefs/fs/littlefs_fs.c:
```diff
@@ -0,0 +1,499 @@
+    if (ret >= 0) {
+        return 0;
+    }
+    else {
```
else not needed, due to return above

vincent-d replied:
Removed

smlng commented on pkg/littlefs/fs/littlefs_fs.c:
```diff
@@ -0,0 +1,499 @@
+    if (ret >= 0) {
+        return 0;
+    }
+    else {
```
else not needed

vincent-d replied:
Removed

smlng commented on tests/unittests/tests-littlefs/tests-littlefs.c:
```diff
@@ -0,0 +1,403 @@
+    res = vfs_unlink("/test-littlefs/test.txt");
+    TEST_ASSERT_EQUAL_INT(0, res);
+}
+#include <stdio.h>
```
why here?

vincent-d replied:
Debug leftover, removed

smlng commented on tests/unittests/tests-littlefs/tests-littlefs.c:
```diff
@@ -0,0 +1,403 @@
+/*
+ * Copyright (C) 2016 OTA keys S.A.
```
2018 is close, anyway 2017 is the year?

vincent-d replied:
oops, copy paste

smlng commented on sys/include/fs/littlefs_fs.h:
```diff
@@ -0,0 +1,105 @@
+ * @ingroup     pkg_littlefs
+ * @brief       RIOT integration of littlefs
+ *
+ * Integration of littlefs into RIOT
```
sounds like what `@brief` is already telling 😉

vincent-d replied:
Removed

vincent-d commented:
Updated to latest upstream commit, and fixed unittests.

I had to patch for mips to build, so I use my fork and I opened an upstream PR (https://github.com/geky/littlefs/pull/3) to fix the errors.

smlng commented:
I still like to get #7104 merged first (for the fix as discussed above) and rebase this afterwards. Otherwise this needs some reasonable squashing, too.

smlng commented:
#7104 was merged, please rebase and squash

vincent-d commented:
Rebased and squashed!

smlng commented:
Compiled and ran unittest on macOS: works. But with DEBUG output on, I get several errors:

```
/RIOT/pkg/littlefs/fs/littlefs_fs.c:313:50: error: format specifies type 'unsigned int' but the argument has type
      'size_t' (aka 'unsigned long') [-Werror,-Wformat]
          (void *)filp, (void *)fp, (void *)src, nbytes);
                                                 ^~~~~~
/RIOT/core/include/debug.h:95:32: note: expanded from macro 'DEBUG'
#define DEBUG(...) DEBUG_PRINT(__VA_ARGS__)
                               ^~~~~~~~~~~
/RIOT/core/include/debug.h:56:20: note: expanded from macro 'DEBUG_PRINT'
            printf(__VA_ARGS__); \
                   ^~~~~~~~~~~
/RIOT/pkg/littlefs/fs/littlefs_fs.c:329:51: error: format specifies type 'unsigned int' but the argument has type
      'size_t' (aka 'unsigned long') [-Werror,-Wformat]
          (void *)filp, (void *)fp, (void *)dest, nbytes);
                                                  ^~~~~~
/RIOT/core/include/debug.h:95:32: note: expanded from macro 'DEBUG'
#define DEBUG(...) DEBUG_PRINT(__VA_ARGS__)
                               ^~~~~~~~~~~
/RIOT/core/include/debug.h:56:20: note: expanded from macro 'DEBUG_PRINT'
            printf(__VA_ARGS__); \
                   ^~~~~~~~~~~
/RIOT/pkg/littlefs/fs/littlefs_fs.c:345:37: error: format specifies type 'long' but the argument has type 'off_t'
      (aka 'long long') [-Werror,-Wformat]
          (void *)filp, (void *)fp, off, whence);
                                    ^~~
/RIOT/core/include/debug.h:95:32: note: expanded from macro 'DEBUG'
#define DEBUG(...) DEBUG_PRINT(__VA_ARGS__)
                               ^~~~~~~~~~~
/RIOT/core/include/debug.h:56:20: note: expanded from macro 'DEBUG_PRINT'
            printf(__VA_ARGS__); \
                   ^~~~~~~~~~~
/RIOT/pkg/littlefs/fs/littlefs_fs.c:360:19: error: use of undeclared identifier 'filp'
          (void *)filp, (void *)buf);
                  ^
/RIOT/pkg/littlefs/fs/littlefs_fs.c:396:19: error: use of undeclared identifier 'filp'
          (void *)filp, (void *)buf);
                  ^
5 errors generated.

```

vincent-d commented:
Oops, will fix

vincent-d commented:
Should be fixed. Amended directly.

I don't have the first errors on my machine, so it would be great if you could confirm it's fixed.

smlng commented:
re-ACK, works like charm with and w/o Debug on macOS! NICE!

smlng commented:
then let's do this: ACK & GO

benpicco commented on pkg/littlefs/fs/littlefs_fs.c:
```diff
@@ -0,0 +1,497 @@
+    if (ret >= 0) {
+        entry->d_ino = info.type;
+        entry->d_name[0] = '/';
+        strncpy(entry->d_name + 1, info.name, VFS_NAME_MAX - 1);
```
Why is a `/` added in front of every directory entry here?
